OpenRA.Mods.Common/Graphics/DefaultSpriteSequence.cs

```diff
@@ -19,70 +19,51 @@
 
-		public EmbeddedSpritePalette(uint[] filePalette = null, Dictionary<int, uint[]> framePalettes = null)
-		{
-			this.filePalette = filePalette;
-			this.framePalettes = framePalettes;
-		}
+		static readonly MiniYaml NoData = new MiniYaml(null);
 
-		public bool TryGetPaletteForFrame(int frame, out uint[] palette)
+		public DefaultSpriteSequenceLoader(ModData modData)
 		{
-			if (framePalettes == null || !framePalettes.TryGetValue(frame, out palette))
-				palette = filePalette;
+			var metadata = modData.Manifest.Get<SpriteSequenceFormat>().Metadata;
+			if (metadata.TryGetValue("BgraSheetSize", out var yaml))
+				BgraSheetSize = FieldLoader.GetValue<int>("BgraSheetSize", yaml.Value);
 
-			return palette != null;
+			if (metadata.TryGetValue("IndexedSheetSize", out yaml))
+				IndexedSheetSize = FieldLoader.GetValue<int>("IndexedSheetSize", yaml.Value);
 		}
-	}
-
-	public class DefaultSpriteSequenceLoader : ISpriteSequenceLoader
-	{
-		static readonly MiniYaml NoData = new MiniYaml(null);
-
-		public DefaultSpriteSequenceLoader(ModData modData) { }
 
 		public virtual ISpriteSequence CreateSequence(ModData modData, string tileset, SpriteCache cache, string image, string sequence, MiniYaml data, MiniYaml defaults)
 		{
-			return new DefaultSpriteSequence(modData, tileset, cache, this, image, sequence, data, defaults);
+			return new DefaultSpriteSequence(cache, this, image, sequence, data, defaults);
 		}
 
-		public IReadOnlyDictionary<string, ISpriteSequence> ParseSequences(ModData modData, string tileSet, SpriteCache cache, MiniYamlNode imageNode)
+		int ISpriteSequenceLoader.BgraSheetSize => BgraSheetSize;
+		int ISpriteSequenceLoader.IndexedSheetSize => IndexedSheetSize;
+
+		IReadOnlyDictionary<string, ISpriteSequence> ISpriteSequenceLoader.ParseSequences(ModData modData, string tileset, SpriteCache cache, MiniYamlNode imageNode)
 		{
 			var sequences = new Dictionary<string, ISpriteSequence>();
-			MiniYaml defaults;
-			try
-			{
-				var node = imageNode.Value.Nodes.SingleOrDefault(n => n.Key == "Defaults");
-				defaults = node?.Value ?? NoData;
-				imageNode.Value.Nodes.Remove(node);
-			}
-			catch (Exception e)
-			{
-				throw new InvalidDataException($"Error occurred while parsing {imageNode.Key}", e);
-			}
+			var node = imageNode.Value.Nodes.SingleOrDefault(n => n.Key == "Defaults");
+			var defaults = node?.Value ?? NoData;
+			imageNode.Value.Nodes.Remove(node);
 
 			foreach (var sequenceNode in imageNode.Value.Nodes)
 			{
 				using (new Support.PerfTimer($"new Sequence(\"{imageNode.Key}\")", 20))
 				{
 					try
 					{
-						var sequence = CreateSequence(modData, tileSet, cache, imageNode.Key, sequenceNode.Key, sequenceNode.Value, defaults);
+						var sequence = CreateSequence(modData, tileset, cache, imageNode.Key, sequenceNode.Key, sequenceNode.Value, defaults);
+						((DefaultSpriteSequence)sequence).ReserveSprites(modData, tileset, cache, sequenceNode.Value, defaults);
```

Comment:
I'm confused - how is this a safe cast if `CreateSequence` is virtual?

Reply:
This is safe by policy: users that subclass the `*Loader` are expected to also subclass the same `*SpriteSequence`.